### ISSUE TITLE
tf.config.LogicalDeviceConfiguration memory_limit does not really work ?

### ISSUE THREAD

**Issue by fitoule** [tensorflow/tensorflow#53822]
**System information**
- Custom code 
- Linux Ubuntu 20.04 (tested also 18.04)
- Computer (not mobile device) 
- TensorFlow installed from pip install:
- TensorFlow version 2.7 (tested also 2.4,2.5,2.6,2.8)
- Python version: 3.7 (tested also 3.8)
- Bazel version (if compiling from source): unbuntu 20.04 no self compil
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 11.2.2/8.2.0.1 
- GPU model and memory: NVIDIA RTX3090 24GB (tested also RTX2060 anX GTX1660) Driver 460.x

Hello I successfully ran the profiler tool on ma classification model to profile the maximum memory usage. Because I want to use different CNN on a same GPU. But I'm really baffled by the results of the profiler. Let me explain  

I have a NVIDIA RTX3090 with 24GB memory so for my small CNN I set 512 memory limit in my code before all use with this code :  
` tf.config.set_logical_device_configuration(gpus[0],[tf.config.LogicalDeviceConfiguration(memory_limit=512)])`

It seems to work because of the tensorflow logs
`2022-01-19 16:24:13.615890: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1525] Created device /job:localhost/replica:0/task:0/device:GPU:0 with **512 MB memory:**  -> device: 0, name: GeForce RTX 3090, pci bus id: 0000:2d:00.0, compute capability: 8.6`

Nvidia-smi shows that GPU use 419MiB are used
![smi first](https://user-images.githubusercontent.com/35316806/150172707-7ff0eef5-6feb-4915-9662-724b16356254.png)

Then I start a batch to make the inference on the classification model with batch size = 1
and tensorboard shows that the model use about 100MiB
![tensorboard](https://user-images.githubusercontent.com/35316806/150173173-823b1989-6174-4b99-a9aa-5fe272ba9b6e.png)

so theoretically I could have set a small memory limit (under 512) but .. here is the real use of the memory given by Nvidia-smi is 1869MiB !
![nvidiaamemoryused](https://user-images.githubusercontent.com/35316806/150173725-d1d8943d-9518-43b2-9607-ef0acfcbc3ec.png)

I tested the code in Tensorflow 2.4,2.5,2.6,2.7 and 2.8. With different CUDA CUDNN but it is the same. Memory limit seems to be applied at Tensorflow level but not at the real GPU memory ( Nvidia level ).
Did I miss something ? It would be very usefull to be able to manage the memory of a model !

**Comment by sushreebarsa** [comment]
@fitoule 
In order to expedite the trouble-shooting process here,Could you please fill the issue [template](https://github.com/tensorflow/tensorflow/issues/new/choose),
Thanks!

**Comment by fitoule** [comment]
Sorry I have added it on the first post

**Comment by sushreebarsa** [comment]
@fitoule 
In order to expedite the trouble-shooting process, please provide a code snippet to reproduce the issue reported here. Thanks!

**Comment by fitoule** [comment]
Ok I've created a notebook that you can download and execute
https://github.com/fitoule/tensorflow_gpu_memory-/blob/main/DemoMemoryIssue.ipynb

**Comment by sushreebarsa** [comment]
@Saduf2019 Was able to replicate the issue on colab using TF [v2.6.0](https://colab.research.google.com/gist/sushreebarsa/cbd73983fec67d8b01c7121967bd5d32/53822-tf-2-6.ipynb),[2.7.0](https://colab.research.google.com/gist/sushreebarsa/3560712e22b6cb42ccf55890559dd189/53822.ipynb#scrollTo=DbMAqcBq8EN_) and [tf-nightly](https://colab.research.google.com/gist/sushreebarsa/44a183a9e0f7da664a62b4fe716b7518/53822.ipynb#scrollTo=831bjfFS-cYi) ,please find the attached gists for reference.Thanks!

**Comment by fitoule** [comment]
I made further investigations. Actually the command line works but documentation is not enough. on my test when I set memory_limit=200. 
A) When I Call import tensorflow => NVIDIA memory allocated id 423MiB
B) When I Call the code with memory limit => NVIDIA memory allocated is 423+200=623MiB
C) When a first inference is called then TensorFlow add a C part memory 938MiB (+423+200)  Total =  1561 MiB

So I understand that A+C is a constant that is needed by TensorFlow and the memory_limit affects only the B part. I tested on many different model.
A+B depends on the driver or GPU HW.

So now it's clear for me. But finally documentation could mention this because see for a low model about 100MiB I need 1.5GB ram, it 's confusing.
